test/e2e/nodeportlocal_test.go

```diff
@@ -0,0 +1,413 @@
+		}
+		rules = append(rules, rule)
+	}
+	checkForNPLRuleInIPTABLES(t, data, r, antreaPod, rules, present)
+}
+
+func checkForNPLRuleInIPTABLES(t *testing.T, data *TestData, r *require.Assertions, antreaPod string, rules []nplRuleData, present bool) {
+	cmd := []string{"iptables", "-t", "nat", "-S"}
+	t.Logf("Verifying iptables rules %v", rules)
+	err := wait.Poll(time.Second, defaultTimeout, func() (bool, error) {
+		stdout, _, err := data.runCommandFromPod(antreaNamespace, antreaPod, agentContainerName, cmd)
+		if err != nil {
+			t.Logf("Error while checking rules in IPTABLES: %v", err)
+			// In case of error, retry
+			return false, nil
+		}
+		for _, rule := range rules {
+			ruleSpec := []string{
+				"-p", "tcp", "-m", "tcp", "--dport",
+				fmt.Sprint(rule.nodePort), "-j", "DNAT", "--to-destination", fmt.Sprintf("%s:%d", rule.podIP, rule.podPort),
+			}
+			rs := strings.Join(append([]string{"-A", "ANTREA-NODE-PORT-LOCAL"}, ruleSpec...), " ")
+			t.Logf("Searching for IPTABLES rule: %v", rs)
+
+			if strings.Contains(stdout, rs) && present {
+				t.Logf("Found rule in IPTABLES")
+			} else if !strings.Contains(stdout, rs) && !present {
+				t.Logf("Rule not found in IPTABLES")
+			} else {
+				return false, nil
+			}
+		}
+		return true, nil
+	})
+	r.NoError(err, "Poll for IPTABLES rules check failed")
+}
+
+func checkTrafficForNPL(t *testing.T, data *TestData, r *require.Assertions, nplAnnotations []k8s.NPLAnnotation, clientName string) {
+	for i := range nplAnnotations {
+		err := data.runNetcatCommandFromTestPod(clientName, nplAnnotations[i].NodeIP, int32(nplAnnotations[i].NodePort))
+		r.NoError(err, "Traffic test failed for NodeIP: %s, NodePort: %d", nplAnnotations[i].NodeIP, nplAnnotations[i].NodePort)
+	}
+}
+
+func enableNPLInConfigmap(t *testing.T, data *TestData) {
+	if err := data.mutateAntreaConfigMap(func(data map[string]string) {
+		antreaAgentConf, _ := data["antrea-agent.conf"]
+		antreaAgentConf = strings.Replace(antreaAgentConf, "#  NodePortLocal: false", "  NodePortLocal: true", 1)
+		antreaAgentConf = strings.Replace(antreaAgentConf, "#nplPortRange: 40000-41000", "nplPortRange: 40000-41000", 1)
+		t.Logf("Updated antreaAgentConf after enabling NodePortLocal: %v", antreaAgentConf)
+		data["antrea-agent.conf"] = antreaAgentConf
+	}, false, true); err != nil {
+		t.Fatalf("Failed to enable NodePortLocal feature: %v", err)
+	}
+}
+
+func updateNPLPortRangeInConfigmap(t *testing.T, data *TestData, oldStartPort, oldEndPort, newStartPort, newEndPort int) {
+	if err := data.mutateAntreaConfigMap(func(data map[string]string) {
+		antreaAgentConf, _ := data["antrea-agent.conf"]
+		oldPortRange := fmt.Sprintf("nplPortRange: %d-%d", oldStartPort, oldEndPort)
+		newPortRange := fmt.Sprintf("nplPortRange: %d-%d", newStartPort, newEndPort)
+		antreaAgentConf = strings.Replace(antreaAgentConf, oldPortRange, newPortRange, 1)
+		t.Logf("Updated antreaAgentConf after updating port range for NodePortLocal: %v", antreaAgentConf)
```

Comment:
please do not print the ConfigMap in the logs. It makes the logs hard to read and does not provide much value.